### DISCUSSION TITLE
Own custom keywords and references in text

### DISCUSSION THREAD

**Discussion by isorokuYam** [vkbo/novelWriter#1399] [Ideas]
Hello.
1. I could miss something from documentation, but as I understand every category has it's own keyword and all other custom categories has to use `@custom` keyword. 
Is there possibility to allow user define own keyword for each category (atleast custom)?
2. Is there possibility to allow user hide reference inside text under some word? similar to html hyperlink `<a href="url">link text</a>`, where user sees "link text", but in fact it's link. Something like 
"He took his `<a href="@object some_object">item name</a>` and blablabla...
Maybe in Markdown it would be [item name](@object some_object), but never used the language before :<

**Comment by vkbo** [comment]
You can make as many folders for each category as you want, and name them whatever you want, but you cannot add new keywords. The way the keywords are defined in the code is fixed. It is possible to add more to the code of course, if there is a demand for extending the list.

When I get around to revamping the tags and references system, I will consider using the label for the category as it is defined on the root folder, instead of the current hard coded label. That would at least make it easier to distinguish them.

As for in-line references, this is not possible and will likely not be added in the foreseeable future. The main reason for this is that they aren't intended to be used in this way. Each reference is made for a given heading, so setting it in the text wouldn't actually change where the reference is made from and links back to. If you do need to associate a specific reference with a specific paragraph for whatever reason, you can set the keyword and reference immediately before the paragraph. Like this:

```markdown
## Some Heading

@char: Bob, John

Some text, bla bla bla.

@char: Jane

Some more text, where Jane appears, bla bla bla.
```

The references to Bob, John and Jane will all point back to "Some Heading" though, and not to their location in the text. Meta data is collected on headings (intended for chapters and scenes really).

**Reply by isorokuYam** [reply]
Thanks for detailed answer.
If I would too add some keywords for myself, is it enough if I will modify constants.py? Just for working preview on click.

**Reply by vkbo** [reply]
There are a few places you will need to add them. They are defined in `nwItemClass` in enum.py. You can add more there. Then make sure you also add them to the lookup dictionaries in constants.py where the others are mentioned: `nwKeyWords` and `nwLabels`. The first is where the `@` keyword is defined, and the labels are just for the GUI.

If you want them to also show up in the Outline View, you need to add columns for them in `nwOutline` in enum.py and add the corresponding labels in `nwLabels` in constants.py.

I think that is all. I try to keep all these things in those two files.
